### PR TITLE
[GPU] Propagate output flag to mutable_data dependencies to ensure correct syncronization

### PR DESCRIPTION
**Pull request by mletavin**
### Details:
 - *When mutable_data is output node, additional output synchronization events should be created for its dependencies in order to correctly wait for all kernels completion before output data can be read*

### Tickets:
 - *73896*
 - *81465*

### CODE REVIEW

File: src/plugins/intel_gpu/src/graph/graph_optimizer/basic_memory_dependencies.cpp [hunk 1]
```diff
@@ -75,7 +76,14 @@ void basic_memory_dependencies::run(program& p) {
         // this output has to land on the primitve restriction list. Otherwise memory reuse can corrupt final results.
         node->add_memory_dependency(past_outputs);
         // if current node is an output add it to the outputs list after restriction.
-        if (node->is_output())
+        if (node->is_output()) {
             past_outputs.push_back(node->id());
+            if (node->is_type<mutable_data>()) {
```
Comment (vladimir-paramuzov) [inline src/plugins/intel_gpu/src/graph/graph_optimizer/basic_memory_dependencies.cpp:81]: please add a test for that.

Comment (mletavin) [reply src/plugins/intel_gpu/src/graph/graph_optimizer/basic_memory_dependencies.cpp:81]: Added top_k_layer_tests.md_sync test

File: src/plugins/intel_gpu/tests/test_cases/arg_max_gpu_test.cpp [hunk 2]
```diff
@@ -1013,3 +1014,2050 @@ TEST(top_k_layer_tests, sort_probabilities_by_indices) {
         EXPECT_EQ(out_buffer[i], ref_vec[i]);
     }
 }
+
+const std::vector<float> input_vec1 = {
+5.4456432e-05
```
Comment (sshlyapn) [inline src/plugins/intel_gpu/tests/test_cases/arg_max_gpu_test.cpp:1019]: In general looks good to me, but do we really need such a huge hardcoded vector? If so, maybe it makes sense to combine values by 8-10 elements per line. Or maybe we can replace these two vectors by single with 'sum' primitive results?

File: src/plugins/intel_gpu/tests/test_cases/arg_max_gpu_test.cpp [hunk 3]
```diff
@@ -1013,3 +1014,2050 @@ TEST(top_k_layer_tests, sort_probabilities_by_indices) {
...
+,3.506987e-05
+,0.00013208817
+,3.4427045e-05
+};
+const int output_ref = 341;
+
+TEST(top_k_layer_tests, md_sync) {
+    static const int32_t x_size = 1, y_size = 1, feature_num = 1001, batch_num = 1;
+    const int top_k = 1;
+    layout inp_l = { data_types::f32, format::yxfb, { batch_num, feature_num, x_size , y_size } };
```
Comment (lznamens) [inline src/plugins/intel_gpu/tests/test_cases/arg_max_gpu_test.cpp:3029]:
```suggestion
    layout inp_l = { data_types::f32, format::yxfb, { batch_num, feature_num, x_size, y_size } };
```

Comment (mletavin) [reply src/plugins/intel_gpu/tests/test_cases/arg_max_gpu_test.cpp]: Fixed

**Review by sshlyapn — APPROVED** [review]

**Comment by evgenytalanin-intel** [comment]
Failing pip-conflicts was an issue resolved later, merging without passing this check

**Pull request merged**